### ISSUE TITLE
Rule enhancement: "key-spacing" allow aligned and non-aligned style per group

### ISSUE THREAD

**Issue by doberkofler** [eslint/eslint#1728]
I would be great to relax the "key-spacing" rule by allowing to use type aligned AND the non-aligned style per group. Currently every group must be either aligned or not and it would be helpful to only have each group to be either aligned or not.

Example (`"key-spacing": [2, {"align": "value", "beforeColon": false,"afterColon": true}]`):
```
var obj1 = {
	prop: 'value',
	p:    'value'
};

var obj2 = {
	prop: 'value',
	p: 'value'
};

8:4  error  Missing space before value for key "p"  key-spacing
```

A possible way to configure this would be to have an additional option "align-optional" that has the same functionality as currently align has but is only only considered per group.

**Comment by nzakas** [comment]
Can you explain in more detail? I'm having trouble understanding what you're requesting. It helps if you can phase you description as what you're trying to do, what ESLint is outputting when you try it, and what you would like it to do. Ik now you were trying to provide some of that info above, it's just not very clear to me.

**Comment by doberkofler** [comment]
In the above example, I've used the option "align": "value" and because the values in obj1 are properly aligned no error is reported. In obj2 the values are not aligned and therefore an error is reported.

Instead of requiring all blocks (the documentation of the rule uses this term for a block of properties not separated by a newline) to be formatted the same, it would be helpful to allow a block to be formatted either by aligning or by using whitespace.

My idea on how to define this would be introduce a new option (align-optional) that allows a block to be either aligned or using whitespace.

**Comment by nzakas** [comment]
Okay, I think what you're asking for is a "consistent" option that says all properties in a block must be consistent with the other properties.

**Comment by doberkofler** [comment]
Not being a native English speaker, I seem to reach my limits. Please let me try once more:
I'm asking for a new (more relaxed) option of align that allows a block to be either aligned or to be using whitespace, so each block is consistent in itself but different blocks can be styled differently.

**Comment by nzakas** [comment]
Yup, that's exactly what I'm saying.

**Comment by doberkofler** [comment]
Excellent, but why is the issue is still marked as question.

**Comment by mourner** [comment]
:+1: looking for this too. I had a different solution in mind though — do not relax the default `key-spacing` behavior, but instead introduce a new option that would indicate that either one should be enforced in each individual property group.

**Comment by nzakas** [comment]
I think we're saying the same thing. `value: "consistent"` would let you have different spacing per block but would still expect one of the two existing options.

**Comment by mourner** [comment]
@nzakas yep, I'd be happy with that too!

**Comment by btmills** [comment]
Is it desirable to allow some groups to be aligned on colon and others on value within the same file, or is it reasonable to require "alignment of property groups is optional, but if a group is aligned, it should be aligned on (colon|value, pick just one)"?

I ask because implementing optional yet consistent alignment in a single mode wouldn't be too bad, but detection and enforcement across value and colon mode would get pretty hairy.

**Comment by doberkofler** [comment]
@btmills In my opinion picking just one (colon or value) would be just fine and would actually not want them to be mixed. Are you planning to work in this?

**Comment by mourner** [comment]
> "alignment of property groups is optional, but if a group is aligned, it should be aligned on (colon|value, pick just one)"

Sounds perfect.

**Comment by mourner** [comment]
Also note that pretty often, it makes sense to align properties differently within one object literal. Example from my code: https://github.com/mapbox/pbf/blob/master/index.js#L340

I think value/colon alignment should be checked only within blocks separated by new lines regardless of actual code structure (whether they're part of the same object literal or not).

**Comment by nzakas** [comment]
Isn't that what we do already?

**Comment by btmills** [comment]
```js
/// Supported: in align mode, groups may be aligned to different columns
var obj = {
    firstProperty:  1,
    secondProperty: 2,

    third:          function() {
        return 3;
    },
    fourthProperty: function() {
        return 4;
    }
};

// Not yet supported: some groups aligned, some groups not aligned
var obj = {
    firstProperty:  1,
    secondProperty: 2,

    third: function() {
        return 3;
    },
    fourthProperty: function() {
        return 4;
    }
};
```

**Comment by btmills** [comment]
Idea:

Right now, a group is defined as a collection of sequential Property AST nodes, where each successive node begins on the line immediately following the preceding node. A blank line between properties separates groups.

Let's redefine a group as a collection of sequential Property AST nodes, all having their keys on adjacent lines. Lines between keys separate groups. So in the example above, `firstProperty` and `secondProperty`, having their keys on adjacent lines, would form a group. However, the keys for `third` and `fourthProperty` are separated by multiple lines, so they would not be considered a group.

The rationale for this is that vertical alignment makes sense when the properties are visually one block. When you break up the keys, as is the case above with the body of each function separating the keys, vertical alignment is harder to see and therefore unnecessary.

I'm going to submit a PR to see what this would look like. It should solve this issue.

**Comment by btmills** [comment]
Nevermind, I got excited and didn't consider the original request. :flushed: However, if someone comes along looking for what I described above, I have a working, tested, documented implementation!

**Comment by btmills** [comment]
> "alignment of property groups is optional, but if a group is aligned, it should be aligned on (colon|value, pick just one)"

With this goal in mind, what would such an option look like? I don't think we want it to be optional by default - presumably someone might want to require alignment all the time. Here's one idea to get the discussion going, though I'm not sure how much I like it:

```js
[2, {
    "align": "value",
    "requireAlignment": false // Defaults to true
}]
```

**Comment by mourner** [comment]
@btmills agree with your thoughts on alignment. I'd make the optional alignment the default though — I've never ever seen code with _all_ properties aligned, it's just not practical.

**Comment by doberkofler** [comment]
@mourner My original proposal matched the `requireAlignment` option defaulting to true as proposed by @btmills because I did not want to change the default behavior. I personally would actually prefer `requireAlignment` to have a default of false.

**Comment by IanVS** [comment]
Maybe I'm misunderstanding something here, but what is the benefit of having an alignment rule like `value` or `colon` and then pulling out its teeth with a `requireAlignment` that defaults to false?  If you don't want to require alignment, it seems you shouldn't specify `align`.  The example given in the original description looks inconsistent and exactly the kind of thing which using an alignment rule is intended to avoid.  The only real usefulness I can imagine is what @btmills suggested in https://github.com/eslint/eslint/issues/1728#issuecomment-72918511. What am I missing?

**Comment by doberkofler** [comment]
@IanVS My original intention was to relax this rule and no longer require all groups to be aligned the same way but rather all items in one group to be aligned consistently.

@btmills has created a very nice example for this:
```javascript
/// Supported: in align mode, groups may be aligned to different columns
var obj = {
    firstProperty:  1,
    secondProperty: 2,

    third:          function() {
        return 3;
    },
    fourthProperty: function() {
        return 4;
    }
};

// Not yet supported: some groups aligned, some groups not aligned
var obj = {
    firstProperty:  1,
    secondProperty: 2,

    third: function() {
        return 3;
    },
    fourthProperty: function() {
        return 4;
    }
};
```

**Comment by IanVS** [comment]
It seems that @btmills abandoned his solution because it did not meet your original request.  Are you saying that his proposal to only align if keys are on subsequent lines would be sufficient for your needs?

**Comment by doberkofler** [comment]
@IanVS Yes, if in the above code example also the "Not yet supported" part would be optionally allowed.

**Comment by IanVS** [comment]
If I understand correctly, that "not yet supported" part is what @btmills referred to as having ready to go in https://github.com/eslint/eslint/issues/1728#issuecomment-72918511. Although the code has likely diverged by now. 

It does seem to be useful, though. Any chance of getting it in?

**Comment by nzakas** [comment]
@IanVS if someone wants to take lead and put together a PR, we'll happily review it.

**Comment by IanVS** [comment]
I believe @btmills had something already put together at one point.

**Comment by btmills** [comment]
If I still have that code, it'll be in an unpublished branch on my laptop. I won't have access to it until sometime next week; I'll check then and update here.

**Comment by btmills** [comment]
I didn't have the code anymore, and it likely would have drowned me in merge conflicts anyway, but I reconstructed it and filed #2447.

**Comment by moll** [comment]
I wouldn't mind something akin to https://github.com/eslint/eslint/issues/1728#issuecomment-90690944 either. The commit that closed this issue (https://github.com/eslint/eslint/commit/ee8e0f2937d704fb3835331daec5ebe3f1c59145) solves the alignment per group, but not mixing aligned and non-aligned keys in a single file (or even single object if you want to go fancy).

**Comment by mourner** [comment]
Yeah, I think we should either open a new issue, or reopen this one. Enforcing alignment is kind of meaningless unless you can intermix these styles.

**Comment by doberkofler** [comment]
Obviously (I opened this issue) I'm all for it!

**Comment by mayatron** [comment]
+1 for allowing multiple styles (or adding the `requireAlignment` option)
